handler/datasets.go

```diff
@@ -0,0 +1,193 @@
+	"HH":                         "All Households",
+	"UR_HH":                      "All usual residents in households",
+	"UR":                         "All usual residents",
+}
+
+// handleDatasetDataType handles a kafka message corresponding to Dataset event type
+func (h *ContentPublishedHandler) handleDatasetDataType(ctx context.Context, cpEvent *models.ContentPublished, cfg config.Config) error {
+	datasetID, edition, version, err := getIDsFromURI(cpEvent.URI)
+	if err != nil {
+		log.Error(ctx, "error while attempting to get Ids for dataset, edition and version", err)
+		return err
+	}
+
+	// ID is be a combination of the dataset id and the edition like so: <datasets_id>-<edition>
+	generatedID := fmt.Sprintf("%s-%s", datasetID, edition)
+	logData := log.Data{
+		"uid_generated": generatedID,
+	}
+
+	// Make a call to DatasetAPI
+	datasetMetadataPublished, err := h.DatasetCli.GetVersionMetadata(ctx, "", cfg.ServiceAuthToken, cpEvent.CollectionID, datasetID, edition, version)
+	if err != nil {
+		log.Error(ctx, "cannot get dataset published contents version %s from api", err)
+		return err
+	}
+
+	logData["content_published"] = datasetMetadataPublished
+	log.Info(ctx, "datasetAPI response ", logData)
+
+	var uri string
+	if len(datasetMetadataPublished.DatasetLinks.LatestVersion.URL) > 0 {
+		uri = datasetMetadataPublished.DatasetLinks.LatestVersion.URL
+	} else if len(datasetMetadataPublished.DatasetDetails.Links.Version.URL) > 0 {
+		uri = datasetMetadataPublished.DatasetDetails.Links.Version.URL
+	} else {
+		uri = datasetMetadataPublished.Version.Links.Version.URL
+	}
+
+	parsedURI, err := url.Parse(uri)
+	if err != nil {
+		log.Error(ctx, "error parsing the metadata uri", err)
+		return err
+	}
+
+	// Mapping Json to Avro
+	versionDetails := models.VersionDetails{
+		ReleaseDate: datasetMetadataPublished.ReleaseDate,
+	}
+
+	datasetDetailsData := models.DatasetDetails{
+		Title:          datasetMetadataPublished.Title,
+		Summary:        datasetMetadataPublished.Description,
+		CanonicalTopic: datasetMetadataPublished.CanonicalTopic,
+		Subtopics:      datasetMetadataPublished.Subtopics,
+		Edition:        edition,
+		DatasetID:      datasetID,
+		Type:           "dataset_landing_page",
+		PopulationType: models.PopulationType{},
+		Dimensions:     []models.Dimension{},
+	}
+
+	if err := populateCantabularFields(ctx, datasetMetadataPublished, &datasetDetailsData); err != nil {
+		return fmt.Errorf("failed to populate cantabular fields: %w", err)
+	}
+
+	if datasetMetadataPublished.Keywords != nil {
+		datasetDetailsData.Keywords = *datasetMetadataPublished.Keywords
+	}
+
+	versionMetadata := models.CMDData{
+		UID:            generatedID,
+		URI:            parsedURI.Path,
+		VersionDetails: versionDetails,
+		DatasetDetails: datasetDetailsData,
+	}
+
+	datasetVersionMetadata := models.MapVersionMetadataToSearchDataImport(versionMetadata)
```

Comment:
Suggestion: we could just pass the datasetMetadataPublished and cpEvent objects into this method and extend to handle all transformations

Reply:
I'm restructuring the mapper a bit as well, there is no need to use the intermediate structs, and the mapping will be much more clearer. Doing the following:
1- Create the event in the handler, initialised with values that come directly from the cpEvent (the datasetID, edition, version, etc. )
2- call dataset api to obtain the metadata
3- map the metadata to the corresponding values in search data import event, without any intermediary structs (done in the models package, as mapper)